mina.netty/src/main/java/org/kaazing/mina/netty/bootstrap/ConnectionlessServerBootstrap.java

```diff
@@ -82,80 +88,76 @@ public ChannelHandler getParentHandler() {
 
     private final class ConnectionlessParentChannelHandler extends SimpleChannelUpstreamHandler {
 
-        private final Map<SocketAddress, Map<SocketAddress, Channel>> childChannelsByLocalAddress;
+        // remote address --> child channel
+        private final Map<SocketAddress, NioDatagramChannel> childChannels;
 
-        public ConnectionlessParentChannelHandler() {
-            childChannelsByLocalAddress = new ConcurrentHashMap<>();
+        ConnectionlessParentChannelHandler() {
+            childChannels = new ConcurrentHashMap<>();
         }
 
         @Override
-        public void channelConnected(ChannelHandlerContext ctx,
-                ChannelStateEvent e) throws Exception {
-            // TODO Auto-generated method stub
+        public void channelConnected(ChannelHandlerContext ctx, ChannelStateEvent e) throws Exception {
             super.channelConnected(ctx, e);
         }
 
         @Override
-        public void messageReceived(ChannelHandlerContext ctx, MessageEvent e)
-                throws Exception {
-
-//            // lookup child channel based on local and remote addresses
-//            Channel channel = e.getChannel();
-//            SocketAddress remoteAddress = e.getRemoteAddress();
-//            Channel childChannel = getChildChannel(channel, remoteAddress, true);
-//
-//            // deliver message received to child channel pipeline
-//            fireMessageReceived(childChannel, e);
+        public void childChannelOpen(ChannelHandlerContext ctx, ChildChannelStateEvent e) throws Exception {
+            ((IoAcceptorChannelHandler) parentHandler).childChannelOpen(ctx, e);
+            NioDatagramChannel childChannel = (NioDatagramChannel) e.getChildChannel();
+            childChannel.getWorker().executeInIoThread(
+                    () -> fireChannelConnected(childChannel, childChannel.getRemoteAddress())
+            );
         }
 
         @Override
-        public void exceptionCaught(ChannelHandlerContext ctx, ExceptionEvent e)
-                throws Exception {
+        public void childChannelClosed(ChannelHandlerContext ctx, ChildChannelStateEvent e) throws Exception {
+            Channel childChannel = e.getChildChannel();
+            childChannels.remove(childChannel.getRemoteAddress());
+        }
+
+        @Override
+        public void messageReceived(ChannelHandlerContext ctx, MessageEvent e) throws Exception {
+            // lookup child channel based on local and remote addresses
+            Channel channel = e.getChannel();
+            NioDatagramChannel childChannel = getChildChannel(channel, e.getRemoteAddress());
+
+            // deliver message received to child channel pipeline
+            childChannel.getWorker().executeInIoThread(
+                    () -> fireMessageReceived(childChannel, e.getMessage())
+            );
```

Comment:
It doesn't seem to create object for this. Same lambda is being used for multiple calls.